From b79d6671dcc616c26a92f36ae19b16f057f7bbff Mon Sep 17 00:00:00 2001
From: Jiri Pirko <jpirko@redhat.com>
Date: Fri, 27 Jun 2014 14:48:35 -0400
Subject: [net] update __dev_notify_flags() to send rtnl msg

Message-id: <1403880516-18932-1-git-send-email-jpirko@redhat.com>
Patchwork-id: 85058
O-Subject: [patch rhel7 net 1/2] net: dev: update __dev_notify_flags() to send rtnl msg
Bugzilla: 1113925
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

BZ1113925
https://brewweb.devel.redhat.com/taskinfo?taskID=7633144
Tested on my testing machine

upstream commit a528c219df2e865e178c538c7178961dfed5a13c
Author: Nicolas Dichtel <nicolas.dichtel@6wind.com>
Date:   Wed Sep 25 12:02:44 2013 +0200

    dev: update __dev_notify_flags() to send rtnl msg

    This patch only prepares the next one, there is no functional change.
    Now, __dev_notify_flags() can also be used to notify flags changes via
    rtnetlink.

Signed-off-by: Jiri Pirko <jpirko@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 73eb8cd..165f096 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -2376,7 +2376,9 @@ extern int  dev_ethtool(struct net *net, struct ifreq *);
 extern unsigned int dev_get_flags(const struct net_device *);
 extern int  __dev_change_flags(struct net_device *, unsigned int flags);
 extern int  dev_change_flags(struct net_device *, unsigned int);
-extern void  __dev_notify_flags(struct net_device *, unsigned int old_flags);
+void   __dev_notify_flags(struct net_device *,
+        unsigned int old_flags,
+        unsigned int gchanges);
 extern int  dev_change_name(struct net_device *, const char *);
 extern int  dev_set_alias(struct net_device *, const char *, size_t);
 extern int  dev_change_net_namespace(struct net_device *,
diff --git a/net/core/dev.c b/net/core/dev.c
index 026897e..3fd6f40 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4813,10 +4813,14 @@ int __dev_change_flags(struct net_device *dev, unsigned int flags)
  return ret;
 }
 
-void __dev_notify_flags(struct net_device *dev, unsigned int old_flags)
+void __dev_notify_flags(struct net_device *dev, unsigned int old_flags,
+   unsigned int gchanges)
 {
  unsigned int changes = dev->flags ^ old_flags;
 
+ if (gchanges)
+  rtmsg_ifinfo(RTM_NEWLINK, dev, gchanges);
+
  if (changes & IFF_UP) {
   if (dev->flags & IFF_UP)
    call_netdevice_notifiers(NETDEV_UP, dev);
@@ -4847,10 +4851,7 @@ int dev_change_flags(struct net_device *dev, unsigned int flags)
   return ret;
 
  changes = old_flags ^ dev->flags;
- if (changes)
-  rtmsg_ifinfo(RTM_NEWLINK, dev, changes);
-
- __dev_notify_flags(dev, old_flags);
+ __dev_notify_flags(dev, old_flags, changes);
  return ret;
 }
 EXPORT_SYMBOL(dev_change_flags);
diff --git a/net/core/rtnetlink.c b/net/core/rtnetlink.c
index 188722b..cbcaba4 100644
--- a/net/core/rtnetlink.c
+++ b/net/core/rtnetlink.c
@@ -1665,9 +1665,8 @@ int rtnl_configure_link(struct net_device *dev, const struct ifinfomsg *ifm)
  }
 
  dev->rtnl_link_state = RTNL_LINK_INITIALIZED;
- rtmsg_ifinfo(RTM_NEWLINK, dev, ~0U);
 
- __dev_notify_flags(dev, old_flags);
+ __dev_notify_flags(dev, old_flags, ~0U);
  return 0;
 }
 EXPORT_SYMBOL(rtnl_configure_link);
-- 
1.7.1